fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectory.java

```diff
@@ -0,0 +1,383 @@
+     * @param name name for the file reference
+     * @param reference the file reference being inserted
+     */
+    public void writeFDBLuceneFileReference(final String name, final FDBLuceneFileReference reference) {
+        LOG.trace("writeFDBLuceneFileReference {}", reference);
+        hasWritten = true;
+        txn.set(metaSubspace.pack(name), reference.getTuple().pack());
+        fileReferenceCache.put(name, reference);
+    }
+
+    /**
+     * Writes data to the given block under the given id.
+     * @param id id for the data
+     * @param block block for the data to be stored in
+     * @param value the data to be stored
+     */
+    public void writeData(long id, int block, byte[] value) {
+        LOG.trace("writeData id={}, block={}, valueSize={}", id, block, value.length);
+        hasWritten = true;
+        txn.set(dataSubspace.pack(Tuple.from(id, block)), value);
+    }
+
+    /**
+     * Seeks known data from the directory.
+     * @param resourceDescription the name of the resource (file)
+     * @param referenceFuture the reference where the data supposedly lives
+     * @param block the block where the data is stored
+     * @return Completable future of the data returned
+     * @throws IOException if blockCache fails to get the data from the block
+     * @throws NullPointerException if a reference with that id hasn't been written yet.
+     */
+    public CompletableFuture<byte[]> seekData(String resourceDescription, CompletableFuture<FDBLuceneFileReference> referenceFuture, int block) throws IOException {
+        try {
+            LOG.trace("seekData resourceDescription={}, block={}", resourceDescription, block);
+            final FDBLuceneFileReference reference = referenceFuture.join(); // Tried to fully pipeline this but the reality is that this is mostly cached after listAll, delete, etc.
+            Long id = reference.getId();
+            return blockCache.get(Pair.of(id, block),
+                    () -> {
+                        Long value = reads.getOrDefault(resourceDescription + ":" + id, 0L);
+                        value += 1;
+                        reads.put(resourceDescription + ":" + id, value);
+                    return txn.get(dataSubspace.pack(Tuple.from(id, block)));
+                }
+            );
+        } catch (ExecutionException e) {
+            throw new IOException(e);
+        }
+    }
+
+    /**
+     * Lists all references in the subspace. Puts all references in the cache.
+     * Logs the count of references, and the total size of the data.
+     * All references are put into the cache because each composite file will be required to serve a query and the file references are small.
+     *
+     * @return String list of names of lucene file references
+     */
+    @Override
+    public String[] listAll() {
+        List<String> outList = new ArrayList<>();
+        List<String> displayList = new ArrayList<>();
```

Comment:
AFAICT, `displayList` is only used for debug logging purposes, so maybe we don't want to create a new array if `LOG.isDebugEnabled()!=true`?

Reply:
Fixing...